File: guests/swift/nav.swift
// The nav scene, Swift port — guests/rust/nav.rs, tools/scenes/nav.steps.

import Foundation

let DETAIL: UInt64 = 7
let SETTINGS: UInt64 = 8

let app = KayaApp()

app.build { tx in
    tx.window(title: "nav")
    let status = tx.signal(.str("at root"))
    let root = tx.column { root in
        tx.label(bind: status)  // label#0
        tx.button(
            "open detail",
            onClick: { inner in  // button#0
            inner.pushEntry(
                DETAIL, title: "detail",
                onPopped: { tx2 in tx2.write(status, .str("popped detail")) })
            let pane = inner.column { pane in
                let caption = inner.signal(.str("detail pane"))
                inner.label(bind: caption)
                return pane
            }
            inner.mountIn(DETAIL, pane)
            inner.write(status, .str("pushed detail"))
            })
        tx.button(
            "open settings",
            onClick: { inner in  // button#1
            // Nothing has popped, so the handler agrees and confirms.
            inner.pushEntry(
                SETTINGS, title: "settings", interceptBack: true,
                onBackRequested: { tx2 in
                    tx2.write(status, .str("back requested"))
                    tx2.popEntry()
                })
            let pane = inner.column { pane in
                let caption = inner.signal(.str("settings pane"))
                inner.label(bind: caption)
                return pane
            }
            inner.mountIn(SETTINGS, pane)
            inner.write(status, .str("pushed settings"))
            })
        return root
    }
    tx.mount(root)
}

app.run()

File: guests/ocaml/nav.ml
(* The nav scene, OCaml port — guests/rust/nav.rs, tools/scenes/nav.steps. *)

open Kaya_app

let detail = 7L
let settings = 8L

let () =
  let app = Kaya_app.create () in

  build app (fun () ->
     window ~title:"nav" ();
     let s = signal Scalar.Str ("at root") in
     let on_detail () =
       push_entry ~title:"detail"
         ~on_popped:(fun () -> write s ("popped detail"))
         detail;
       (let caption = signal Scalar.Str ("detail pane") in
        let pane = column [ label ~bind:caption ] () in
        mount_in detail pane;
        write s ("pushed detail"))
        
     in
     let on_settings () =
       push_entry ~title:"settings" ~intercept_back:true
         ~on_back_requested:(fun () ->
           write s ("back requested");
           pop_entry ())
         settings;
       (let caption = signal Scalar.Str ("settings pane") in
        let pane = column [ label ~bind:caption ] () in
        mount_in settings pane;
        write s ("pushed settings"))
        
     in
     let root =
       column
         [
           label ~bind:s (* label#0 *);
           button ~text:"open detail" ~on_click:on_detail;
           button ~text:"open settings" ~on_click:on_settings;
         ]
         ()
     in
     mount root);

  exit (run app)
